### PR TITLE
Add Replay Buffer Events and Requests

### PR DESCRIPTION
**Pull request by lsdaniel**
Documentation is included, not really much, but I needed events for the replay buffer starting and stopping. These events have only been available as of 17.0.0 (I think). I don't think that the OBS API has an event for saving the replay yet, so that can't happen yet :/.

### CODE REVIEW

**Comment by lsdaniel** [comment]
I have found requests to start/stop the replay buffer, but I haven't found one to flush the replay buffer to disk. Anyone have any ideas?

**Comment by Palakis** [comment]
Thanks for this, that is good quality code. I've fixed conflicts caused by my recent refactor of the codebase, along with some style nitpicks.

> I have found requests to start/stop the replay buffer, but I haven't found one to flush the replay buffer to disk. Anyone have any ideas?

I'll have a look at OBS' API and come back to you.

**Comment by Palakis** [comment]
Looks like even OBS itself doesn't have anything to flush a replay buffer. Maybe deleting or emptying the file will do the trick. Is it what you want to achieve, @lsdaniel?

**Comment by lsdaniel** [comment]
@Palakis Sorry for wording it that way, I meant save the current replay buffer (what is in memory) to the disk. This is normally achieved by using a hotkey, but there is no front-end function to do it using the libobs api. Unless I am wrong, which I may be. I filed a bug on [OBS's bug tracker](https://obsproject.com/mantis/view.php?id=955), but have heard nothing so far. The farthest I've understood how the replay buffer works is by looking at [this function](https://github.com/jp9000/obs-studio/blob/cfe77fdaaeacb78027e8c045739ead7a6f673302/plugins/obs-ffmpeg/obs-ffmpeg-mux.c#L685) in the obs code.

**Comment by Palakis** [comment]
AFAIK the replay buffer is constantly being written to while active, so no need to trigger a manual save or something.

**Comment by Palakis** [comment]
The replay buffer actually being a file.

**Comment by XBase-xOR** [comment]
OBS doesn't have any GUI element that can trigger writing the file. Pressing a hotkey that is configured [here](https://x-base.org/img/OBS_replay_buffer.png) (red border) is the only way.

The replay buffer in memory is constantly updated in memory and not written to a file unless you trigger to write it to a file by using that hotkey.

I also updated #90 for further clarification.

**Comment by Palakis** [comment]
Thanks @XBase-xOR. I'm not a regular user of this feature so I didn't have all the details.
I can't find this "Save replay buffer" hotkey, both in 19.0.3 and the latest OBS master codebase. Is there something special to enable?

**Comment by Palakis** [comment]
@lsdaniel I managed to manually trigger buffer saves through the hotkey system!

**Comment by Palakis** [comment]
@XBase-xOR nevermind, If found the option to enable the Replay buffer and the hotkey now shows up.

**Comment by XBase-xOR** [comment]
Ah OK, I didn't even know that OBS hides it if the replay buffer is disabled.

**Comment by Palakis** [comment]
Things left to do:
- Return an error if the replay buffer is disabled in OBS' settings
- Handle the case where no hotkey is bound to save the replay buffer to disk: bind a dummy hotkey? return an error?

**Comment by TStod** [comment]
It is probably worth noting that the replay buffer is only enabled when using the `Simple` output mode with the `Enable Replay Buffer` checkbox selected.

Hotkey settings for capturing/enabling/disabling replay are stored but hidden if the above settings are not set/enabled.

File: WSRequestHandler.cpp [hunk 1]
```diff
@@ -1638,7 +1736,7 @@ void WSRequestHandler::HandleSetRecordingFolder(WSRequestHandler* req) {
 /**
  * Get the path of  the current recording folder.
  *
- * @return {Stsring} `rec-folder` Path of the recording folder.
+ * @return {String} `rec-folder` Path of the recording folder.
```
Comment (haganbmj) [inline WSRequestHandler.cpp]: Thanks for fixing these. Sloppy typos.

Comment (Palakis) [inline WSRequestHandler.cpp]: Shit happens 😄

**Comment by XBase-xOR** [comment]
@Palakis Thanks so much dude, changed my tool from using hotkey to the new API method and it works like a charm!

**Comment by ihrishix** [comment]
@Palakis @haganbmj Is there any way i can remotely enable the Replay Buffer from obs settings and set the duration ? (Obs settings > output > replay buffer)

**Pull request merged**